data/lib/index.js

```diff
@@ -0,0 +1,139 @@
+import { pipe } from 'it-pipe'
+import { batch } from 'streaming-iterables'
+import { CID } from 'multiformats/cid'
+
+import { MAX_BATCH_GET_ITEMS } from '../tables/constants.js'
+import { createCarTable } from '../tables/car.js'
+import { createFerryTable } from '../tables/ferry.js'
+import { createAggregateService } from './aggregate-service.js'
+
+/**
+ * @typedef {import('../types.js').FerryOpts} FerryOpts
+ * @typedef {import('../types.js').CarItem} CarItem
+ * @typedef {import('../types.js').CarItemFerry} CarItemFerry
+ *
+ * @typedef {object} FerryCtx
+ * @property {string} region
+ * @property {string} tableName
+ * @property {FerryOpts} [options]
+ * 
+ * @typedef {object} CarTableCtx
+ * @property {string} region
+ * @property {string} tableName
+ * @property {import('../types.js').CarOpts} [options]
+ */
+
+/**
+ * Add cars to a loading ferry.
+ *
+ * @param {CarItemFerry[]} cars
+ * @param {FerryCtx} ferryCtx
+ */
+export async function addCarsToFerry (cars, ferryCtx) {
+  const ferryTable = createFerryTable(ferryCtx.region, ferryCtx.tableName, ferryCtx.options)
+
+  const ferryId = await ferryTable.getFerryLoading()
+  await ferryTable.addCargo(ferryId, cars)
+
+  return {
+    id: ferryId
+  }
+}
+
+/**
+ * Sets current Ferry as ready if not previously done
+ *
+ * @param {string} ferryId
+ * @param {FerryCtx} ferryCtx
+ */
+export async function setFerryAsReady (ferryId, ferryCtx) {
+  const ferryTable = createFerryTable(ferryCtx.region, ferryCtx.tableName, ferryCtx.options)
+
+  // Update state of ferry to ready
+  try {
+    await ferryTable.setAsReady(ferryId)
+  } catch (/** @type {any} */ error) {
+    // If error is for condition we can safely ignore it given this was changed in a concurrent operation
+    if (error.name !== 'ConditionalCheckFailedException') {
+      throw error
+    }
+  }
+}
+
+/**
+ * Sets current Ferry offer.
+ *
+ * @param {string} ferryId
+ * @param {object} ctx
+ * @param {FerryCtx} ctx.car
+ * @param {FerryCtx} ctx.ferry
+ * @param {import('../types.js').StorefrontSignerCtx} ctx.storefront
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} ctx.aggregationServiceConnection
+ */
+export async function setFerryOffer (ferryId, ctx) {
```

Comment:
Note that how this is implemented, there are guarantees that if (sadly is possible) dynamoDB stream handler is called multiple times for same thing, that two different offers won't be created.